### ISSUE TITLE
rs_currents product problem in extractEnv

### ISSUE THREAD

**Issue by ecologistpablo** [IMOS-AnimalTracking/remora#30]
**Describe the bug**
When utilising extractEnv (which still isn't working on windows) the function works with data up until 2021. I dug into the back-end of the function again and found the product being utilised is only available from 1993 - 2020 (https://thredds.aodn.org.au/thredds/catalog/IMOS/OceanCurrent/GSLA/DM/catalog.html). There is another product available from 2021 - 2023, the build_urls script just needs updating to include an ifelse loop to say if dates < 2020.12.31 then utilise different product (NRT which is available from 2011 - 2023: https://thredds.aodn.org.au/thredds/catalog/IMOS/OceanCurrent/GSLA/NRT/catalog.html) in  the ocean current section.

I've included steps to reproduce the error-code and output of working/non working workflow if you wish to replicate:

**To Reproduce**
> #extractEnv rs_currents problem
>   #13.07.23
> 
> #We have a detections file from IMOS for all Grey Nurse Shark detections along the east coast of Aus
> #To demonstrate the problem, let's download 1 week of data in 2020 which will work
> #and one week in 2021 which won't as the NRT product in the thredds database
> 
> det <- read_csv("Inputs/IMOS_detections.csv")
Rows: 780125 Columns: 33                                                                                                                          
── Column specification ───────────────────────────────────────────────────────────────────────────────────────
Delimiter: ","
chr  (14): transmitter_id, tagging_project_name, species_common_name, species_scientific_name, animal_sex, ...
dbl  (14): tag_id, transmitter_deployment_id, CAAB_species_id, WORMS_species_aphia_id, receiver_id, receive...
lgl   (2): transmitter_dual_sensor, embargo_date
dttm  (3): detection_datetime, detection_corrected_datetime, transmitter_deployment_datetime

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
> det$detection_datetime <- as.Date(det$detection_datetime)
> 
> #lets try 2021 first
> det1 <- det %>% 
+   filter(detection_datetime > as.Date("2021-01-01"))
> 
> #the broken error code
> det1 <- 
+   extractEnv(df = det1,
+              X = "receiver_deployment_longitude", 
+              Y = "receiver_deployment_latitude", 
+              datetime = "detection_datetime", 
+              env_var = "rs_current",
+              cache_layers = T,
+              crop_layers = TRUE,
+              full_timeperiod = FALSE,
+              fill_gaps = T,
+              folder_name = "test",
+              .parallel = T)
Extracting environmental data only on days detections were present; between 2021-01-03 and 2023-01-31 (311 days)
This may take a little while...
Accessing and downloading IMOS environmental variable: rs_current
Finding IMOS Ocean Current data...
Error in map(.x, .f, ...) : ℹ In index: 1.
ℹ With name: 2021.
Caused by error in `open.connection()`:
! HTTP error 404.

Downloaded layers are cached in the `imos.cache` folder in your working directory
Extracting and appending environmental data
Filling gaps in environmental data by extracting median values from a 20km buffer around detections that fall on 'NA' values
Error in h(simpleError(msg, call)) : 
  error in evaluating the argument 'x' in selecting a method for function 'extract': object 'env_stack' not found
> #now 2020
> det1 <- det %>% 
+   filter(detection_datetime > as.Date("2020-01-01"))
> #which works
> det1 <- 
+   extractEnv(df = det1 [1:7,], #only one week
+              X = "receiver_deployment_longitude", 
+              Y = "receiver_deployment_latitude", 
+              datetime = "detection_datetime", 
+              env_var = "rs_current",
+              cache_layers = F,
+              crop_layers = TRUE,
+              full_timeperiod = FALSE,
+              fill_gaps = T,
+              folder_name = "test",
+              .parallel = T)
Extracting environmental data only on days detections were present; between 2020-02-05 and 2020-03-25 (2 days)
This may take a little while...
Accessing and downloading IMOS environmental variable: rs_current
Finding IMOS Ocean Current data...
  |=====================================================================================================| 100%Extracting and appending environmental data
Filling gaps in environmental data by extracting median values from a 20km buffer around detections that fall on 'NA' values
> #lets check
> colnames(det1)
 [1] "detection_datetime"                 "detection_corrected_datetime"      
 [3] "transmitter_id"                     "tag_id"                            
 [5] "transmitter_deployment_id"          "tagging_project_name"              
 [7] "species_common_name"                "species_scientific_name"           
 [9] "CAAB_species_id"                    "WORMS_species_aphia_id"            
[11] "animal_sex"                         "receiver_project_name"             
[13] "installation_name"                  "station_name"                      
[15] "receiver_id"                        "receiver_name"                     
[17] "receiver_deployment_id"             "receiver_deployment_longitude"     
[19] "receiver_deployment_latitude"       "transmitter_sensor_type"           
[21] "transmitter_sensor_raw_value"       "transmitter_sensor_unit"           
[23] "transmitter_sensor_slope"           "transmitter_sensor_intercept"      
[25] "transmitter_type"                   "transmitter_serial_number"         
[27] "transmitter_estimated_battery_life" "transmitter_status"                
[29] "transmitter_deployment_longitude"   "transmitter_deployment_latitude"   
[31] "transmitter_deployment_datetime"    "transmitter_dual_sensor"           
[33] "embargo_date"                       "rs_gsla"                           
[35] "rs_vcur"                            "rs_ucur"                           
[37] "rs_current_velocity"                "rs_current_bearing"                
> #and it worked

**Desktop (please complete the following information):**
 - MacOS Big Surf Version 11.7.B
 - Browser: chrome
 - Version: RStudio 4.3.1 (2023-06-16) -- "Beagle Scouts" 
 - _REMORA:_ 0.7.1 (downloaded 2023-07-13)

**Additional context**

[IMOS_detections_dec2020-Jan2021.csv](https://github.com/IMOS-AnimalTracking/remora/files/12035239/IMOS_detections_dec2020-Jan2021.csv)

**Comment by ecologistpablo** [comment]
@ianjonsen @vinayudyawer

**Comment by ianjonsen** [comment]
Thanks, I will update this in the next 24-48 h. The Windows issue requires more work.

**Comment by ecologistpablo** [comment]
Thank you for your assistance and quick reply, it's truly appreciated (not fussed about windows as I have access to macs fortunately)! It will be a bit more complex than I originally thought to approach inputting the correct file-string into the function as it varies across four options due to timestamps (I believe) in the string. Instead of it only being (across multiple days): IMOS_OceanCurrent_HV_20210108**T000000Z**_GSLA_FV02_NRT.nc it could be any one of the four options:
IMOS_OceanCurrent_HV_20210108**T000000Z**_GSLA_FV02_NRT.nc
IMOS_OceanCurrent_HV_20210108**T060000Z**_GSLA_FV02_NRT.nc
IMOS_OceanCurrent_HV_20210108**T120000Z**_GSLA_FV02_NRT.nc
IMOS_OceanCurrent_HV_20210108**T180000Z**_GSLA_FV02_NRT.nc

So on top of .build_urls needing updating to provide the new url to find the nc files the file string formula in .pull_env (I think that's where they are but am not 100% sure) will need updating too. I can't seem to crack it myself and push an update to you sorry, so good luck and I hope this has helped a little bit.

**Comment by ianjonsen** [comment]
Yes, there's a chain of updates that are needed. Step 1 - I have updated the code to handle cases like your example where some or all years of IMOS Current data are not available. `extractEnv` will now exit gracefully, returning a data.frame with either: i) current variables when IMOS Current data exists for all years; ii) current variables when IMOS Current data exist for some years and NA's for years with no data; or iii) the original input data when no IMOS Current data exists for any of the years. This is now available on the Staging branch. However, there is an unresolved issue when running extractEnv in parallel, so I would only set `.parallel = FALSE` for now.

Step 2 - adapt code to optionally provide the NRT Current data when DM Current does not exist. I don't want to provide NRT Current data automatically, ie. without the user realizing it, as the NRT Current data has not been subjected to the same quality control scrutiny as the DM Current data. I will make this change soon.

**Comment by ecologistpablo** [comment]
Thanks for the update, I thought there would be variability in the product so it's understandable that you want people to make a conscious decision regarding choosing it. For me it's just better than nothing or attempting to access other data products (Like BRAN) the long way so it would help greatly having access to NRT!

**Comment by ianjonsen** [comment]
Access to BRAN is coming shortly

**Comment by ecologistpablo** [comment]
I saw that pull, wonderful news! If I want to try access these latest updates in the staging branch and re-run the function offline without remora using the code of it, would I only need **build_urls**, **extract_var** and **pull_var** as functions to build extractEnv again to use for a dataset from 2013 - 2023? Is it ready to use or do I need to wait for it to be pushed into master and finalised? Thanks :)

**Comment by ianjonsen** [comment]
Yes, those are the functions you would need, though I don’t see why you would want to run them independent of the package? Do you want the values appended to the detections data frame summarized in some other way? 

Yes, you can use from the staging branch. Those and a few other updates have not been widely tested enough to merge into the master branch, but they are otherwise stable. You can install this version via: `install_github(IMOS-AnimalTracking/remora@staging)`

**Comment by ecologistpablo** [comment]
Ah, thanks for the installation code!! I wasn't aware you can install branches that are in other stages (relatively new to github and git) so I thought the only way of accessing them was to re-create independently! I'll give it a go :)
